docs: update the CONTRIBUTING.md (#230176)
AdguardTeam/AdguardFilters: CONTRIBUTING.md

## Workflow for submitting changes

1. Create a new branch for your changes. Please use the following naming convention:
   - `{prefix}/{description}`
   - `{prefix}/{issue-number}-{description}`
   - `{prefix}/{third-party-repo-prefix}-{issue-number}-{description}`

>    Where `{prefix}` is `fix` for breakage/unblocking fixes or `upd` for filter updates/additions.

Reviewer: Both are for fixing issues and adding rules. `upd` - if you forgot to add some changes in fix commit (not complete fix) or changes is not a complete feature (minor changes in docs, for example). For minor improvements.

Author: I see. But I'm not sure how to revise this sentence. Could you share the draft so it aligns with your intention?

Author: I revised it 👀 

>   Where `{prefix}` is:
>   - `fix` — complete fixes or rule additions
>   - `upd` — minor improvements or follow-ups (e.g. a missing change from a prior commit, minor doc edits)

## Workflow for submitting changes

1. Create a new branch for your changes. Please use the following naming convention:
   - `fix-#{issue-number}-{website-domain}` — for breakage or unblocking fixes
   - `upd-#{issue-number}-{website-domain}` — for filter updates or additions

>    Example: `fix-#12345-example.com`

Reviewer: Better leave current format

`{prefixed-branch}/{description}` 
`{prefixed-branch}/{issue-number}-{description}`
`{prefixed-branch}/{third-party-repo-prefix}-{issue-number}-{description}`

```
fix/12345
fix/12345-example.com
fix/ubo-12345-example.com
upd/12345
```

## Workflow for submitting changes

1. Create a new branch for your changes. Please use the following naming convention:
   - `{prefix}/{description}`
   - `{prefix}/{issue-number}-{description}`
   - `{prefix}/{third-party-repo-prefix}-{issue-number}-{description}`

   Where `{prefix}` is:
   - `fix` — complete fixes or rule additions

>    - `upd` — minor improvements or follow-ups (e.g. a missing change from a prior commit, minor doc edits)

Reviewer: better without "or follow-ups"

```suggestion
   - `upd` — minor improvements (e.g. addressing omissions from previous fix commit or making small documentation or comment edits)
```

   If AGLint passes, your PR will be reviewed by a maintainer.
1. If the review is successful, your changes will be merged into the `master` branch.

### Rule placement

> - New rules go at the **top of the relevant section** in the filter file.

Reviewer: Let's slightly change the wording

```suggestion
- By default, new rules should be placed at the top of the relevant section in the filter file.
```

  ```diff
  !NOTE: Regular rules

  + example.org##.banner-ads
  ||example.com/ads/*
  ```

- If the new rule shares the same domain as an existing group of rules,

>   insert it within that group in ascending order, rather than at the top of the section.

Reviewer: ```suggestion
- If the new rule shares the same domain with an existing group of rules,
insert it **within that group** in ascending order by domain name, rather than at the top of the section.
```

Maybe without `in ascending order by domain name`?

> 
>   ```diff
>   ||example.org/images/ads/*
>   + ||example.org/scripts/tracking.js
>   example.org##.example-ads

Reviewer: Most groups have a description/comment (added `! Example group`)

```suggestion
  ```diff
  ! Example group
  ||example.org/images/ads/*
  + ||example.org/scripts/tracking.js
  example.org##.example-ads
```

  ```

- For groups of mixed-type rules sharing the same domains,

>   add the new domain either at the **beginning or at the end**.

Reviewer: ```suggestion
  add the new domain either at the **beginning** or **at the end**.
```

>   This makes it easy to visually compare the lists if they become out of sync.

Reviewer: Add an example with mixed types.

```suggestion
  This makes it easy to visually compare the lists if they become out of sync:
  
 ```diff
-example.com,example.org###ads
+example.com,example.org,test.com###ads
-||example.com^$domain=example.com|example.org
+||example.com^$domain=example.com|example.org|test.com
-||ads.com^$domain=example.com|example.org
+||ads.com^$domain=example.com|example.org|test.com
```
```

  For element hiding rules – at the beginning of the rule:

  ```diff
  - example.com,example.org###ads
  + test.com,example.com,example.org###ads
  ```

  In general, for rules with `|` separator, add a new domain at the end:

  ```diff
  - adserver$domain=example.com|example.org
  + adserver$domain=example.com|example.org|test.com
  ```

  **Avoid:** inserting the domain at a random position in the middle of the list.

### Commit message format

When closing an issue with a commit:

```text
Fix #ISSUE_NUMBER example.org optional description
```

When updating without closing (e.g. asking the user to verify):

```text
Upd #ISSUE_NUMBER example.org description
```

> When using `Upd`, also mark the issue as `A: Waiting for data`.

Reviewer: Regular users cannot change labels.
```suggestion
```

   If AGLint passes, your PR will be reviewed by a maintainer.
1. If the review is successful, your changes will be merged into the `master` branch.

### Rule placement

- By default, new rules should be placed at the top of the relevant section in the filter file.

  ```diff
  !NOTE: Regular rules

  + example.org##.banner-ads
  ||example.com/ads/*
  ```

- If the new rule shares the same domain with an existing group of rules,
  insert it **within that group** in ascending order by domain name, rather than at the top of the section.

  ```diff
  ! Example group
  ||example.org/images/ads/*
  + ||example.org/scripts/tracking.js
  example.org##.example-ads
  ```

- For groups of mixed-type rules sharing the same domains,
  add the new domain either at the **beginning** or **at the end**.
  This makes it easy to visually compare the lists if they become out of sync:

  ```diff
  - example.com,example.org###ads
  + test.com,example.com,example.org###ads
  - adserver.com$domain=example.com|example.org
  + adserver.com$domain=example.com|example.org|test.com

>   ```

Reviewer: Not what I suggested - for the same domain lists you added the the beginning AND at the end.

```
-example.com,example.org###ads
+example.com,example.org,test.com###ads
-||example.com^$domain=example.com|example.org
+||example.com^$domain=example.com|example.org|test.com
-||ads.com^$domain=example.com|example.org
+||ads.com^$domain=example.com|example.org|test.com
```

In other words, if there is a set of rules for a number of sites, their order should be preserved. The only difference is the separator (`,` or `|`)

Author: > If there is a set of rules for a number of sites, their order should be preserved.

Got it. But why did you initially recommend placing the new domain at the end in the first rule? 

Because it is wrong:

```diff
-example.com,example.org###ads
+example.com,example.org,test.com###ads
```

It is correct what I know.

```diff
-example.com,example.org###ads
+test.com,example.com,example.org###ads
```

Reviewer: That rules - example of one group of mixed-type rules.

look:

1. `example.com,example.org,test.com`
2. `example.com|example.org|test.com`
3. `example.com|example.org|test.com`

Domains order is the same. It's immediately clear that the list contains no missing items. And if you add domains in a consistent manner, the list will be easier to read.

Reviewer: > But why did you initially recommend placing the new domain at the end in the first rule?

for non-mixed-type single rules or groups (only cosmetic, or only URL blocking, for example).

   If AGLint passes, your PR will be reviewed by a maintainer.
1. If the review is successful, your changes will be merged into the `master` branch.

### Rule placement

- By default, new rules should be placed at the top of the relevant section in the filter file.

  ```diff
  !NOTE: Regular rules

  + example.org##.banner-ads
  ||example.com/ads/*
  ```

- If the new rule shares the same domain with an existing group of rules,

>   insert it **within that group** in ascending order by domain name, rather than at the top of the section.

Reviewer: ```suggestion
- If the rule applies to a site for which a rules group exists, insert it **within that group**, rather than at the top of the section.
```

  ```diff
  ! Example group
  ||example.org/images/ads/*
  + ||example.org/scripts/tracking.js
  example.org##.example-ads
  ```

> - For rules that share the same rule expression,

Reviewer: ```suggestion
- For rules that share the same rule expression, depending on the type of rule,
```

  add the new domain to the existing expression — either at the **beginning** or **at the end**,
  rather than creating a new rule on a separate line.

>   This makes it easy to add a new domain consistently, treating them as a single rule.

Reviewer: ```suggestion
  This makes it easy to add a new domain consistently, treating them as a single rule.
  
  **Important:** Don't insert a new domain at a random position in the middle of the list.
```

>   For element hiding rules – at the beginning of the rule:

Reviewer: ```suggestion
  * For cosmetic rules – at the beginning of the rule:
```

  ```diff
  - example.com,example.org###ads
  + test.com,example.com,example.org###ads
  ```

>   In general, for rules with `|` separator, add a new domain at the end:

Reviewer: ```suggestion
  * In general, for rules with `|` domains separator, add a new domain at the end:
```

  ```diff
  - adserver.com$domain=example.com|example.org
  + adserver.com$domain=example.com|example.org|test.com
  ```

>   When the group contains both types, keep domain order consistent across all rules.

Reviewer: ```suggestion
  * When the group contains both types, keep domain order consistent across all rules.
```

  This is because only the separator differs (`,` vs `|`) —

>   consistent ordering makes it immediately clear that no domains are missing.

Reviewer: ```suggestion
    This is because only the separator differs (`,` vs `|`) —
    consistent ordering makes it immediately clear that no domains are missing.
```

  ```diff
  -example.com,example.org###ads
  +example.com,example.org,test.com###ads
  -||example.com^$domain=example.com|example.org
  +||example.com^$domain=example.com|example.org|test.com
  -||ads.com^$domain=example.com|example.org
  +||ads.com^$domain=example.com|example.org|test.com
  ```

>   **Avoid:** inserting the domain at a random position in the middle of the list.

Reviewer: ```suggestion
```
moved above

### Commit message format

> When closing an issue with a commit:

Reviewer: ```suggestion
- To close an associated issue, use:
```

```text
Fix #ISSUE_NUMBER example.org optional description
```

> When updating without closing (e.g. asking the user to verify):

Reviewer: ```suggestion
- For updates that do not require closing the issue (e.g. problematic issue, an issue with multiple reports, making changes while solving the problem together with the user)
```

```text
Upd #ISSUE_NUMBER example.org description
```

> 

Reviewer: ```suggestion

If a commit is associated with an issue that has more than one report, a link to the comment can be used as the description.

```
